examples/sealed-auctions/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+            let auction = Auction {
+                item_id: item_id,
+                bids: HashSet::new(),
+                reserve: reserve,
+                max_bid: 0,
+                realized: 0,
+                seller: ctx.sender(),
+                buyer: UserId::default(),
+            };
+            self.auctions.insert(item_id, auction.clone());
+
+            // emit an event for the start of this auction
+            Event::emit(&artifact.base);
+
+            Ok(auction)
+        }
+
+        /// Place a bid on an item
+        /// Anyone other than the seller can place a bid as long as they are not
+        /// on the blacklist
+        pub fn place_bid(&mut self, ctx: &Context, item_id: ItemId, value: usize) -> Result<()> {
+            if !self.auctions.contains_key(&item_id) {
+                return Err(failure::format_err!(
+                    "There is no active auction for item {}.",
+                    item_id
+                ));
+            }
+            let mut auction: &mut Auction = self.auctions.get_mut(&item_id).unwrap();
+            if auction.seller == ctx.sender() {
+                return Err(failure::format_err!(
+                    "The seller cannot also be a bidder for an item."
+                ));
+            }
+            if self.blacklist.contains(&ctx.sender()) {
+                return Err(failure::format_err!(
+                    "You don't have permission to bid. Please contact the auction owner."
+                ));
+            }
+            if value < auction.reserve {
+                return Err(failure::format_err!(
+                    "The bid has to be higher than the reserve of {}.",
+                    auction.reserve
+                ));
+            }
+            if value < auction.max_bid {
+                return Err(failure::format_err!(
+                    "The bid is lower than the current maximum."
+                ));
+            }
+
+            // Create/update bid
+            if !auction.bids.contains(&ctx.sender()) {
+                auction.bids.insert(Bid {
+                    bidder: ctx.sender(),
+                    value: value,
+                });
+            } else {
+                let mut bid = auction.bids.take(&ctx.sender()).unwrap();
+                if bid.value > value || value < auction.max_bid {
+                    return Err(failure::format_err!(
+                        "A new bid has to be greater than {}.",
+                        std::cmp::max(bid.value, auction.max_bid)
+                    ));
+                }
+                bid.value = value;
+                auction.bids.insert(bid);
+            }
+            auction.max_bid = value;
+
+            Ok(())
+        }
+
+        /// Close an auction
+        /// Only the seller can close an auction
+        pub fn close_auction(&mut self, ctx: &Context, item_id: ItemId) -> Result<Auction> {
+            if !self.auctions.contains_key(&item_id) {
```

Comment:
`Entry`

Reply:
Done.

Reply:
Ah! I can't use Entry here. We actually want to remove the auction from the auctions map. So, we first check containment and then subsequently remove it and not get it unlike in the other cases above.

Reply:
`match` on [`take`](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.take), then perhaps?